### PR TITLE
Report flag json extension

### PR DESCRIPTION
**Pull request by EwanMoshi**
# Description:

Please provide details for the change or fix:
- extended the **-report** parameter so it supports outputting in json format.
   - Using **-report=json** will print the report in json format
   - Can also use **-report=html** for the current html format
   - Defaults to html if nothing specified, i.e. all existing projects using **-report** will continue working as if nothing changed
- Report class is now a base class (interface) for HtmlReport and JsonReport. Each subclass is responsible for the generation

# Checklist:

The pull request:
- [x] **Is created against the Dev branch**
- [x] **Is self-contained**
- [ ] **Compiles on Windows, OSX and Linux** (only compiled on windows)
- [ ] **Has accompanying tests**
- [x] **Passes existing tests**
- [x] **Keeps Windows, OSX and Linux at parity**
- [x] **Follows the code style**
- [x] **Includes documentation**

### CODE REVIEW

File: Code/Tools/FBuild/FBuildCore/Helpers/HtmlReport.cpp [hunk 1]
```diff
@@ -0,0 +1,851 @@
+// HtmlReport
```
Comment (EwanMoshi) [inline Code/Tools/FBuild/FBuildCore/Helpers/HtmlReport.cpp:1]: This file is essentially the old `Report.cpp` file but without the common functionality - those have been extracted into the the new `Report.cpp`.

**Comment by EwanMoshi** [comment]
Example JSON output:

```json
{
	"Overview": {
		"cmd line options": "\"fbuild\" -vs -report=json -dist -monitor -cache -clean \"All-x64-Debug\" -wrapperintermediate -wrapperfinal",
		"Target(s)": "All-x64-Debug",
		"Result": "OK",
		"Time": "3.473s",
		"CPU Time": "1m 20.610s (23.2:1)",
		"Remote CPU Time": "0.000s (0.0:1)",
		"Version": "v1.08 Windows",
		"Report Generated": "0.010s  - Fri 20-Jan-2023 - 16:01:38"
	},
	"CPU Time by Node Type": {
		"Object": {
				"Time (s)": 78.697,
				"Processed": 245,
				"Built": 241,
				"Cache Hits": 4,
				"Percentage": 97.6
		},
		"Exe": {
				"Time (s)": 1.895,
				"Processed": 4,
				"Built": 4,
				"Cache Hits": "-",
				"Percentage": 2.4
		},
		"Directory": {
				"Time (s)": 0.012,
				"Processed": 8,
				"Built": 8,
				"Cache Hits": "-",
				"Percentage": 0.0
		},
		"Unity": {
				"Time (s)": 0.005,
				"Processed": 8,
				"Built": 8,
				"Cache Hits": "-",
				"Percentage": 0.0
		},
		"Compiler": {
				"Time (s)": 0.001,
				"Processed": 2,
				"Built": 2,
				"Cache Hits": "-",
				"Percentage": 0.0
		},
		"File": {
				"Time (s)": 0.000,
				"Processed": 615,
				"Built": 274,
				"Cache Hits": "-",
				"Percentage": 0.0
		},
		"Alias": {
				"Time (s)": 0.000,
				"Processed": 1,
				"Built": 1,
				"Cache Hits": "-",
				"Percentage": 0.0
		},
		"ObjectList": {
				"Time (s)": 0.000,
				"Processed": 10,
				"Built": 10,
				"Cache Hits": "-",
				"Percentage": 0.0
		},
		"TextFile": {
				"Time (s)": 0.000,
				"Processed": 4,
				"Built": 4,
				"Cache Hits": "-",
				"Percentage": 0.0
		}
	},
	"Cache Stats": {
		"summary": {
			"Uncacheable": {
				"Time (s)": 0.000,
				"Percentage": 98.4
			},
			"Cache Miss": {
				"Time (s)": 0.000,
				"Percentage": 0.0
			},
			"Cache Hit": {
				"Time (s)": 0.000,
				"Percentage": 1.6
			}
		},
		"details": [
			{
				"Library": "FBuildCore-Lib-x64-Debug",
				"Items": 117,
				"Out-of-Date": {
					"Count": 117,
					"Percentage": 100.0
				},
				"Cacheable": {
					"Count": 0,
					"Percentage": 0.0
				},
				"Hits": {
					"Count": 0,
					"Percentage": 0.0
				},
				"Misses": {
					"Count": 0,
					"Percentage": 0.0
				},
				"Stores": 0,
				"Store Time (s)": 0.000
			},
			....
			{
				"Library": "FBuildWorker-Res-x64-Debug",
				"Items": 1,
				"Out-of-Date": {
					"Count": 1,
					"Percentage": 100.0
				},
				"Cacheable": {
					"Count": 0,
					"Percentage": 0.0
				},
				"Hits": {
					"Count": 0,
					"Percentage": 0.0
				},
				"Misses": {
					"Count": 0,
					"Percentage": 0.0
				},
				"Stores": 0,
				"Store Time (s)": 0.000
			}
		 ]
	},
	"CPU Time by Library": [
		{
			"Time (s)": 36.108,
			"Percentage": 45.9,
			"Obj Built": 117,
			"Type": "ObjectList",
			"Name": "FBuildCore-Lib-x64-Debug"
		},
		....
		{
			"Time (s)": 0.021,
			"Percentage": 0.0,
			"Obj Built": 1,
			"Type": "ObjectList",
			"Name": "FBuildWorker-Res-x64-Debug"
		}
	 ],
	"CPU Time by Item": [
		{
			"Time (s)": 1.366,
			"Type": "Object",
			"Cache": "N/A",
			"Name": "C:\\fastbuild\\tmp\\x64-Debug\\Tools\\FBuild\\FBuildTest\\Tests\\TestObject.obj"
		},
                ....
		{
			"Time (s)": 0.001,
			"Type": "Unity",
			"Cache": "N/A",
			"Name": "Core-Unity-x64-Debug"
		}
	 ],
	"Includes": [
		{
			"Library name": "FBuildCore-Lib-x64-Debug",
			"Includes": [
				{
					"Objects": 117,
					"Included": 117,
					"PCH": "no",
					"Name": "C:\\fastbuild\\Code\\Core\\Containers\\Forward.h"
				},
                                ....
			] 
		},
                ....
                {
			"Library name": "FBuildWorker-Res-x64-Debug",
			"Includes": []
		}
	]
}

File: Code/Tools/FBuild/FBuildCore/Helpers/FBuildStats.cpp [hunk 2]
```diff
@@ -78,9 +78,12 @@ void FBuildStats::OnBuildStop( const NodeGraph & nodeGraph, Node * node )
...
         {
-            Report r;
-            r.Generate( nodeGraph, *this );
-            r.Save();
+            Report* report = options.GetReport();
+
+            report->Generate( nodeGraph, *this );
+            report->Save();
+
+            FREE(report);
```
Comment (ffulin) [inline Code/Tools/FBuild/FBuildCore/Helpers/FBuildStats.cpp:86]: FNEW should be matched with DELETE (FREE is paired with ALLOC)

**Comment by ffulin** [comment]
Looks like there are a few build failures with this change. They look like simple fixes.

**Comment by ffulin** [comment]
Seems like some new compiler errors have been introduced (or perhaps uncovered by the newest fixes)

**Comment by ffulin** [comment]
Thanks for the changes. They have been integrate and will be in the next release (v1.09).

I made some minor style fixes in a followup commit and changed the report behavior such that specifying an invalid type is an error, consistent with other options. (https://github.com/fastbuild/fastbuild/commit/fa9529e8732366568a44e57c177c859bf98a1187)

**Pull request merged**